File: math/238.ProductofArrayExceptSelf.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        size = len(nums)
        output = [1] * size
        left = 1
        for x in range(size - 1):
            left *= nums[x]
            output[x + 1] *= left
        right = 1
        for x in range(size - 1, 0, -1):
            right *= nums[x]
            output[x - 1] *= right
        return output


if __name__ == "__main__":
    # Input:  [1,2,3,4]
    # Output: [24,12,8,6]
    nums = [1, 2, 3, 4]
    s = Solution()
    print(s.productExceptSelf(nums))

File: math/238.ProductofArrayExceptSelf.go
package main

import "fmt"

func productExceptSelf(nums []int) []int {
	res := make([]int, len(nums))
	for i, _ := range res {
		res[i] = 1
	}
	left := 1
	for i := 0; i < len(nums)-1; i++ {
		left *= nums[i]
		res[i+1] *= left
	}
	right := 1
	for i := len(nums) - 1; i > 0; i-- {
		right *= nums[i]
		res[i-1] *= right
	}
	return res
}

func main() {
	nums := []int{1, 2, 3, 4}
	fmt.Println(productExceptSelf(nums))
}
